Sources/App/Core/PackageContributors.swift

```diff
@@ -0,0 +1,132 @@
+    ///   - gitCacheDirectoryPath: path to the cache directory where the clone of the package is stored
+    ///   - packageID: the UUID of the package
+    /// - Returns: PackageAuthors
+    static func extract(gitCacheDirectoryPath: String, packageID: UUID?) throws -> PackageAuthors {
+        let contributorsHistory = try GitHistoryLoader.loadContributorsHistory(gitCacheDirectoryPath: gitCacheDirectoryPath, packageID: packageID)
+        let authors = primaryContributors(candidates: contributorsHistory, threshold: 0.6)
+
+        return PackageAuthors(authors: authors.map { Author(name: $0.name) },
+                              numberOfContributors: contributorsHistory.count - authors.count)
+    }
+
+
+    struct Contributor {
+        /// Total number of commits
+        let numberOfCommits: Int
+        let name: String
+    }
+
+    /// Loads the contributors history from a Git repository
+    struct GitHistoryLoader {
+
+        static func loadContributorsHistory(gitCacheDirectoryPath: String, packageID: UUID?) throws -> [Contributor] {
+            do {
+                let commitHistory = try queryGitHistory(gitCacheDirectoryPath: gitCacheDirectoryPath, packageID: packageID)
+                return try parseGitHistory(log: commitHistory)
+            } catch {
+                throw AppError.analysisError(packageID, "loadContributorsHistory failed: \(error.localizedDescription)")
+            }
+        }
+
+        /// Gets the git history in a string log
+        private static func queryGitHistory(gitCacheDirectoryPath: String, packageID: UUID?) throws -> String {
+
+            if !Current.fileManager.fileExists(atPath: gitCacheDirectoryPath) {
+                throw AppError.cacheDirectoryDoesNotExist(packageID, gitCacheDirectoryPath)
+            }
+
+            // attempt to shortlog
+            do {
+                return try Current.git.shortlog(gitCacheDirectoryPath)
+            } catch {
+                throw AppError.shellCommandFailed("gitShortlog",
+                                                  gitCacheDirectoryPath,
+                                                  "queryGitHistory failed: \(error.localizedDescription)")
+            }
+        }
+
+        /// Parses the result of queryGitHistory into a collection of contributors
+        private static func parseGitHistory(log: String) throws -> [Contributor] {
+            var committers = [Contributor]()
+
+            for line in log.components(separatedBy: .newlines) {
+                let log = line.split(whereSeparator: { $0 == " " || $0 == "\t"})
+
+                if (log.count > 2) {
+                    var identifier = [String]()
+                    for i in 1..<(log.count - 1) {
+                        identifier.append(String(log[i]))
+                    }
+                    var numberOfCommits = 0
+                    if (log.first != nil) {
+                        numberOfCommits = Int(log.first!) ?? 0
+                    }
```

Comment:
I know it's been mentioned above that we need to improve the safety of the line here (the less exclamation marks the better!)

Given with the guard/if statement we know there is at least 3 items in the array. We could maybe do something like:

```swift
let numberOfCommits = Int(log.removeFirst()) ?? 0
let identifier = log.dropLast().joined(separator: " ")
let committer = Contributor ...
```

Docs: https://developer.apple.com/documentation/swift/array/removefirst(_:)
Docs: https://developer.apple.com/documentation/swift/array/droplast(_:)

This would remove the need for `for` loops and array builders!

Reply:
That's interesting! Thank you for bringing this array methods up. This will indeed help removing the foor loops :) I have some questions though:

1. Why after guarding the array to have at least 3 elements is not safe to use `array.first!`, but is safe to call `array.removeFirst()` 🤔  is it because is making an entire copy of the array except the first one and then replacing the entire array with the copy (I noticed it has complexity $O(n)$)? In addition, It is returning the first element (as a non-optional) assuming it was safe from the very beginning, isn't?
2. Is it possible somewhere to see the details of the implementation of this `removeFirst()` or is this hidden?  🥷

Reply:
This comment is less about the safety, and more about the readability and reducing the need for many loops and creating array builders (which requires more memory).

As you say, we already have the guard to make sure there are enough elements so this will be safe whether it's with force unwraps (!) or 'unsafe' functions such as removeFirst.

The implementation is not public 🙂